src/Microsoft.Net.Sdk.AnalyzerRedirecting/RedirectingAnalyzerAssemblyResolver.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using AnalyzerInfo = (int MajorVersion, string PathSuffix, string FullPath);
+
+namespace Microsoft.Net.Sdk.AnalyzerRedirecting;
+
+[Export(typeof(IAnalyzerAssemblyResolver))]
+public sealed class RedirectingAnalyzerAssemblyResolver : IAnalyzerAssemblyResolver
+{
+    private readonly string? _insertedAnalyzersDirectory;
+    private readonly Lazy<ImmutableDictionary<string, List<AnalyzerInfo>>> _analyzerMap;
+
+    [ImportingConstructor]
+    public RedirectingAnalyzerAssemblyResolver()
+        : this(Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "SDK", "RuntimeAnalyzers"))) { }
+
+    // Internal for testing.
+    internal RedirectingAnalyzerAssemblyResolver(string? insertedAnalyzersDirectory)
+    {
+        _insertedAnalyzersDirectory = insertedAnalyzersDirectory;
+        _analyzerMap = new(CreateAnalyzerMap);
+    }
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private ImmutableDictionary<string, List<AnalyzerInfo>> AnalyzerMap => _analyzerMap.Value;
+
+    private ImmutableDictionary<string, List<AnalyzerInfo>> CreateAnalyzerMap()
+    {
+        var builder = ImmutableDictionary.CreateBuilder<string, List<AnalyzerInfo>>(StringComparer.OrdinalIgnoreCase);
+
+        foreach (string topLevelDirectory in Directory.EnumerateDirectories(_insertedAnalyzersDirectory))
+        {
+            foreach (string versionDirectory in Directory.EnumerateDirectories(topLevelDirectory))
+            {
+                foreach (string analyzerPath in Directory.EnumerateFiles(versionDirectory, "*.dll", SearchOption.AllDirectories))
+                {
+                    if (!analyzerPath.StartsWith(versionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string majorVersionStr = version.IndexOf('.') is >= 0 and var index ? version.Substring(0, index) : version;
+                    if (!int.TryParse(majorVersionStr, out int majorVersion))
+                    {
+                        continue;
+                    }
+
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + (EndsWithSlash(versionDirectory) ? 0 : 1));
+                    pathSuffix = Path.GetDirectoryName(pathSuffix);
```

Comment:
I see - so `pathSuffix` here is meant to match exactly the path under the version directory in the product.

So for `"C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll"`
it would be `analyzers\dotnet`
For `"C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\8.0.8\analyzers\dotnet\cs\System.Windows.Forms.Analyzers.CSharp.dll"` it would be `analyzers\dotnet\cs`

It'd be good to have a small writeup of how this works in either code comments or a markdown file.  There's a lot of policy / convention here and it'd be good to spell that out and review it.

Reply:
There is a design doc here: https://github.com/dotnet/sdk/pull/42437. But probably a good idea to replicate some of it in comments directly in the code.

Reply:
Yeah, or link to that design doc.  I think this should work - if we tie the product version to the replacement.

It's a little different for SDK than targeting packs.  For the SDK analyzers there's only ever one version that gets loaded, so it should always be redirected.

For the targeting packs, there might be many loaded, and we only want to hook the one that matches the latest inserted framework version.  For older framework versions, they can continue to load the one from the targeting pack (or we could decide to insert those if we want to NGEN them).  For newer framework versions they don't get replaced.

Reply:
I'm not sure I understand. Are you saying the redirection algorithm should be different somehow?

Currently it works like this:

- in VS we will have for example
  - "C:\Program Files\Microsoft Visual Studio\2022\Preview\SDK\RuntimeAnalyzers\AspNetCoreAnalyzers\8.0.8\analyzers\dotnet\cs\Microsoft.AspNetCore.App.Analyzers.dll"
  - "C:\Program Files\Microsoft Visual Studio\2022\Preview\SDK\RuntimeAnalyzers\WebSDKAnalyzers\8.0.400\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Analyzers.dll"
- a project targets net8.0 and the user has installed SDK 8.0.100 so it tries to load "C:\Program Files\dotnet\packs\Microsoft.AspNetCore.App.Ref\8.0.1\analyzers\dotnet\cs\Microsoft.AspNetCore.App.Analyzers.dll"
  - we will redirect that to the DLL deployed with VS
  - here we are comparing 8.0.400 (deployed) vs 8.0.1 (being loaded), i.e., SDK vs runtime version, I think that's fine since we consider only major and minor components?
- a project uses SDK 8.0.100 and tries to load "C:\Program Files\dotnet\sdk\8.0.100\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Analyzers.dll"
  - we will redirect that to the DLL deployed with VS (8.0.100 vs 8.0.400 are compatible)
- in other cases, if the major/minor versions don't match, we won't redirect anything

Reply:
>  we are comparing 8.0.400 (deployed) vs 8.0.1 (being loaded), i.e., SDK vs runtime version

I don't think comparing SDK versions to framework pack versions is a safe thing to do.  One SDK might bundle multiple targeting pack versions.  Sure it would have the latest targeting pack version, and that's the one we'd want to distribute, but that's coincidental and not a hard rule.  I think it's better to just match the versions from the packs.  So if the analyzer came from a targeting pack versioned 8.0.1, then it gets installed to an 8.0.1 folder.  If the user then tries to load 8.0.n version, it gets replaced with that installed 8.0.1 version.

Base on your other comment [here](https://github.com/dotnet/sdk/pull/42861#discussion_r1758997233) I think you aren't actually comparing SDK versions with framework pack versions, so this might be moot, but that further demonstrates the need for a bit of detail in the design doc or comments about how this replacement works for targeting packs.